dockstore-webservice/src/main/java/io/openapi/api/impl/ToolsApiServiceImpl.java

```diff
@@ -360,45 +359,45 @@ public Response toolsGet(String id, String alias, String toolClass, String descr
-            handleParameter(name, "name", filters);
-            handleParameter(toolname, "toolname", filters);
-            handleParameter(description, "description", filters);
-            handleParameter(author, "author", filters);
-            handleParameter(registry, "registry", filters);
-            handleParameter(String.valueOf(actualLimit), "limit", filters);
-
-            final long numPages = (numEntries.sum()) / actualLimit;
-
-            if (startIndex + actualLimit < numEntries.sum()) {
-                URI nextPageURI = new URI(config.getExternalConfig().getScheme(), null, config.getExternalConfig().getHostname(), port,
-                    ObjectUtils.firstNonNull(config.getExternalConfig().getBasePath(), "") + relativePath,
-                    Joiner.on('&').join(filters) + "&offset=" + (offsetInteger + 1), null).normalize();
-                responseBuilder.header("next_page", nextPageURI.toURL().toString());
-            }
-            URI lastPageURI = new URI(config.getExternalConfig().getScheme(), null, config.getExternalConfig().getHostname(), port,
-                ObjectUtils.firstNonNull(config.getExternalConfig().getBasePath(), "") + relativePath, Joiner.on('&').join(filters) + "&offset=" + numPages, null)
-                .normalize();
-            responseBuilder.header("last_page", lastPageURI.toURL().toString());
+        responseBuilder.header("self_link", createUrlString(scheme, hostname, port, path, query));
+        if (startIndex + actualLimit < numEntries.sum()) {
+            responseBuilder.header("next_page", createUrlString(scheme, hostname, port, path, positionQuery(query, actualLimit, offsetInteger + 1L)));
+        }
+        final long numPages = numEntries.sum() / actualLimit;
+        responseBuilder.header("last_page", createUrlString(scheme, hostname, port, path, positionQuery(query, actualLimit, numPages)));
+
+        return responseBuilder.build();
+    }
 
+    private String createUrlString(String scheme, String hostname, int port, String path, String query) {
+        try {
+            return new URI(scheme, null, hostname, port, path, query, null).normalize().toURL().toString();
         } catch (URISyntaxException | MalformedURLException e) {
-            throw new CustomWebApplicationException("Could not construct page links", HttpStatus.SC_BAD_REQUEST);
+            throw new CustomWebApplicationException("Could not create url string", HttpStatus.SC_BAD_REQUEST);
         }
-        trsListener.loadTRSResponse(hashcode, responseBuilder);
-        return responseBuilder.build();
+    }
+
+    private String positionQuery(String query, long limit, long offset) {
+        List<String> resultNameValues = new ArrayList<>();
+        if (query != null) {
+            for (String nameValue: query.split("&")) {
+                if (!nameValue.startsWith("limit=") && !nameValue.startsWith("offset=")) {
+                    resultNameValues.add(nameValue);
```

Comment:
I think you need to encode the query param value. In looking at it, it looks like that is a bug there already (note the `%26` is an encoded `&` in my example curl; I typed in `abc&def`for the organization value in the Swagger UI). I didn't try running your branch, so maybe it's fixed there, but I suspect not.

```
curl -I GET "https://staging.dockstore.org/api/ga4gh/trs/v2/tools?organization=abc%26def&limit=1000" -H  "accept: application/json"
....
self_link: https://staging.dockstore.org/api/ga4gh/trs/v2/tools?organization=abc&def&limit=1000
last_page: https://staging.dockstore.org/api/ga4gh/trs/v2/tools?organization=abc&def&limit=100&offset=0
```

The page Denis linked to encodes.

Reply:
Good catch, Charles!  I shoulda tested with some weirder values.

For the sake of understanding, in a "form"-type query string, the name and value in each field (name-value pair) is urlencoded https://www.albionresearch.com/tools/urlencode, the resulting encoded values are joined by an unencoded separator '=', and the resulting encoded field values are joined by an unencoded separator '&'.

So, what was happening here is that our code uses `URI.getQuery()` to retrieve the query string, but `getQuery` urldecodes the entire query string.  So, in Charles example, the query string urldecodes to `organization=abc&def&limit=1000`.  Since the escaped ampersand in the org name has been converted to a field separator, any subsequent code is doomed to failure.

`getRawQuery` returns the encoded query string.

Whilst adjusting the code accordingly, I discovered a problem in the `URI` constructor: it insists on encoding the `%` characters in the query string you pass it, which makes it pretty much useless for assembling a url that has a "form"-type query string.  So, I fixed that, too.

Should work now.